### ISSUE TITLE
Can't get ChaiScript lambda that returns bool into C++

### ISSUE THREAD

**Issue by totalgee** [ChaiScript/ChaiScript#481]
* Compiler Used: clang 6.0.0
 * Operating System: Ubuntu 18.04.1 LTS
 * Architecture (ARM/x86/32bit/64bit/etc): x86_64 

### Expected Behavior 

Assuming you've initialized a "chai" object, the following C++ code should print "p(3) = 1" to standard output:

```cpp
{
  // throws bad_any_cast, but should work!
  auto p = chai.eval<std::function<bool (int)>>("fun(x) { x != 0 }");
  std::cout << "p(3) = " << p(3) << '\n';
  // should print: p(3) = 1
}
```

### Actual Behavior 

It throws a `bad_any_cast` error.

Note that similar code with lambdas that return `int` or `void` works fine (see working examples below).

```cpp
{
  auto p = chai.eval<std::function<int (int)>>("fun(x) { x + 10 }");
  std::cout << "p(3) = " << p(3) << '\n';
  // prints: p(3) = 13
}

{
  auto p = chai.eval<std::function<void ()>>("fun() { print(\"nothing to return\") }");
  std::cout << "p(): ";
  p();
  // prints: p() = "nothing to return"
}
```

Note that this used to work (returning lambdas which return bool) in an earlier version of ChaiScript, e.g. 6.1.0 which I'm currently using. I came across this issue when trying to work with the latest (develop or master branch) version.

**Comment by RobLoach** [comment]
``` chaiscript
fun(x) { x + 10 }
```

Don't think this script will run correctly?

1. Use `def` to declare functions
2. `fun` possibly is a keyword?

Try the following instead...
``` cpp
const char *code =
  "def plus10(x) {"
  "  return x + 10"
  "}";
```

Tried it on https://robloach.github.io/ChaiScript.js/ and seemed to run better.
``` chaiscript
def plus10(x) {
    return x + 10
}
print(plus10(10))
```

**Comment by totalgee** [comment]
`fun` is indeed a ChaiScript keyword. The code:

```
fun(x) { x + 10 }
```

Is valid. It creates a lambda ("unnamed function") that takes one argument and returns that argument plus ten...

**Comment by RobLoach** [comment]
Ah, fancy. Didn't know you could do that. I see what you're trying now...
https://github.com/ChaiScript/ChaiScript/blob/develop/cheatsheet.md#sharing-functions

**Comment by ninnghazad** [comment]
Similar issue, the same or probably related: http://discourse.chaiscript.com/t/bad-any-cast-after-update/436

**Comment by totalgee** [comment]
@SG-Skril  Interesting to see progress on this, thanks... I don't know enough about the underlying code to know if that's a valid fix, but your explanation seems to make sense at least. (-;

**Comment by draghan** [comment]
I've made a silly workaround on the client-site by registering global integer constants `True` and `False`:
```
chaiscript::ChaiScript script;
script.add_global_const(chaiscript::const_var(0), "False");
script.add_global_const(chaiscript::const_var(1), "True");
```
...but it's kind of ugly and forces to explicit returning those constants in order to work, which makes it useless in a variety of cases.
I'm glad that there is a proposed solution. :)
@SG-Skril - could you please raise a pull request, so we could trigger any action to fix that bug?

**Comment by totalgee** [comment]
I'd like to close this, but it seems I don't have permission (?) even to close my own issue?

**Comment by totalgee** [comment]
(Resolved by #503)
